### PR TITLE
Add IndexOf/IndexOfT function

### PR DESCRIPTION
**Pull request by elliotcourant**
This just adds the IndexOf function.

### CODE REVIEW

**Comment by ahmetb** [comment]
Honestly, I'm puzzled we didn't have this function. I need some time to figure out why. If I forget to review this,  please gently ping me.

**Comment by elliotcourant** [comment]
Sounds good! Thank you for the fast reply!

**Comment by ahmetb** [comment]
@cleitonmarx what do you think?

**Comment by cleitonmarx** [comment]
@elliotcourant Thanks you for your contribution!

@ahmetb Originally LINQ (Enumerable) doesn't have the method `IndexOf`, just Arrays and Lists types.
We can reproduce the same results using existent LINQ methods: 
```go 
From(slice).TakeWhile(...).Count()-1
```
Here more arguments for the exact same request but in DotNet Core: https://github.com/dotnet/corefx/issues/1415

@elliotcourant Could you describe an use-case for this function?

**Comment by elliotcourant** [comment]
@cleitonmarx I'm using it in a database sharding application as a simple way to loop through a list to see if a column with the given name exists. Its an easy way to see if it exists and if it doesn't handling that, and if it does just grabbing that index from the array. 

It also serves as an easy way to update that item in the array since I have the index instead of the object.

Here is an example: 
https://github.com/elliotcourant/noahdb/blob/aca8f978a56d7c175f809c06675c002817e68bfe/pkg/sql/create.go#L271

**Comment by elliotcourant** [comment]
And at the very least this would just be the shorthand for

```
From(slice).TakeWhile(...).Count()-1
```
But _might_ be more efficient since the TakeWhile adding those items to a collection as it goes and then counting that collection in the end. But the IndexOf is only iterating over the collection and counting.

The difference is probably insignificant for most use cases as far as performance goes.

**Comment by cleitonmarx** [comment]
> But might be more efficient since the TakeWhile adding those items to a collection as it goes and then counting that collection in the end.

TakeWhile doesn't create a new collection, and It just passes the items to the next function while the predicate is false (using the Iterator pattern 😄)

> It also serves as an easy way to update that item in the array since I have the index instead of the object.

This use-case seems a good reason to accept the PR @ahmetb. Thoughts?

@elliotcourant It would be nice to add at least one usage example for better documentation -> example_test.go

File: index.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package linq
+
+// IndexOf returns the index of the first item in the slice to match the predicate.
```
Comment (cleitonmarx) [inline index.go]:
```suggestion
// IndexOf searches for an element that matches the conditions defined by a specified predicate, and returns the zero-based index of the first occurrence within the collection. This method returns -1 if an item that matches the conditions is not found.
```

**Comment by ahmetb** [comment]
I suppose we don’t normally add methods that aren’t part of .NET LINQ. But then it has proper IndexOf method on arrays and Go falls short here. I am ok with it.

**Comment by ECourant** [comment]
I'll update the documentation comments and add an example tonight!

File: index.go (hunk 1 above)
Comment (elliotcourant) [reply index.go]: Added this in 👍

**Comment by elliotcourant** [comment]
Just updated the stuff and added an example, and re-squashed my commits. 👍

**Comment by ahmetb** [comment]
Looks like there are a couple golint + gofmt issues.

**Comment by elliotcourant** [comment]
Yeah my casing was wrong and my indentation was off. Should be fixed now!

**Comment by ahmetb** [comment]
I don't see any issues. I'll have @cleitonmarx have another look and merge. Thanks!

**Comment by elliotcourant** [comment]
Thank you very much you guys, love the library!

**Pull request merged**